### DISCUSSION TITLE
Documentation generation

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#36] [General]
At some point (real soon!), I will sit down and document both the source and the project.

Been looking at documentation generation options. Anyone have experience with these packages?

https://github.com/orgs/cl-doc-systems/repositories?type=all

https://lisp-journey.gitlab.io/blog/overview-of-documentation-generators/

**Comment by kaveh808** [comment]
I have begun writing some documentation for the demos. I split off a file called `/test/demo-kernel.lisp` and have cleaned up, numbered, and documented the examples in there.

Please have a look and test it out if you have time. Let me know what needs clarification and if anything doesn't work.

**Comment by Harag** [comment]
I have recently started using https://diataxis.fr/ for my documentation it helps.

**Comment by kaveh808** [comment]
I am finding the website confusing. Is this a theoretical framework for documentation, or is it an actual tool that reads documentation and generates HTML or something?

What I am thinking would be useful is something that allows us to document the source files by writing descriptions in certain sections and then having that parsed and documentation generated. Ideally something that allows for some sort of markup language and diagramming.

**Reply by Harag** [reply]
It's a theoretical framework. Something to consider when you start thinking about documentation. When generating documentation and writing documentation in source, keeping such a framework in mind is of value.

You could, for instance, take https://github.com/Shinmera/multilang-documentation and, with a hack or two, produce consistently structured documentation that aligns with the framework.

**Comment by Harag** [comment]
As far as the documentation options in the links you have in your original post https://github.com/scymtym/staple looks comprehensive and very customizable.

**Reply by kaveh808** [reply]
I wish there was an example showing what documentation generated what final formatting somewhere on the site.

**Reply by JMC-design** [reply]
https://shinmera.github.io/staple/ any of Shinmera's documentation shows you staples result.

**Comment by lukego** [comment]
I have been impressed with [MGL-PAX](https://github.com/melisgl/mgl-pax) based documentation. But it was all written by the same person so this might only reflect that Gábor Melis is a talented and disciplined writer :-)

Example output: https://github.com/melisgl/mgl#x-28MGL-BP-3A-40MGL-BP-20MGL-PAX-3ASECTION-29
generated from source: https://github.com/melisgl/mgl/blob/master/src/backprop.lisp

**Comment by lukego** [comment]
(Aside: Blender's poor documentation in general, and poor API documentation in particular, seems like one of its biggest weaknesses from the hacker perspective. Even the available books on the topic seem to just be quickly tossed together.)

**Reply by kaveh808** [reply]
From what I gather, they don't even have a C++ API. Only Python.

**Comment by JMC-design** [comment]
I'm probably the only one in the Lisp community that thinks that documentation shouldn't be html but lisp so you can display it however you want, including inside your app.

**Reply by kaveh808** [reply]
HTML does need to be one of the output formats, because it is so widely expected. But the system that parses the documentation should output various formats, including within our system. It should be one of the reflection capabilities of the software.

**Comment by mikelevins** [comment]
So I've worked in production documentation a bunch as a tech writer and as a programmer, and even built publication toolchains for a couple of companies. So I know stuff that might be useful. But it might not. Documentation is a big, complicated area with more nuances than you can shake a stick at.

You might just want to pick one of the Lisp doc generators and call it a day. But I thought I should mention that I've done this kind of work, in case it turns out to be useful.

One thing I did with Lisp before was set up a Docbook toolchain that generated a product's manuals and online help. I used Lisp as one tool in the doc-generation toolchain, for generating some content and for checking links. But I wouldn't want to do that the same way for this project. It's most likely too complicated and overkill.

Clozure Associates created a documentation tool (I didn't write that one; I think Gail Zacharias did). I don't remember how applicable it might be outside their use case, but I'll take a look at it and see.

**Reply by mikelevins** [reply]
I took a look at ccldoc. It works and it's easy to get working under sbcl. Some folks have mentioned preferring a Lispy syntax, and ccldoc's syntax satisfies that requirement, but as a writer, it's not a syntax I would prefer to work with on a daily basis.

Here's an example from ccldoc's repo (the toplevel document about ccldoc itself):

https://github.com/Clozure/ccldoc/blob/master/ccldoc.ccldoc

In case folks do like it, here's all it took to make it work on sbcl 2.2.6 on my macOS Monterey system:

1. clone ccl-compat to a convenient location
    https://github.com/phoe/ccl-compat
2. add "(in-package :asdf)" to the top of ccl-compat.asd
3. load ccl-compat.asd and then (asdf:load-system :ccl-compat)
4. clone ccldoc to a convenient location
    https://github.com/Clozure/ccldoc
5. (load "source/ccldoc.asd") and then (asdf:load-system :ccldoc)

Then I can do

`(defparameter *d* (ccldoc:load-document "/path/to/ccldoc/ccldoc.ccldoc"))`

(where "/path/to/ccldoc/" is replaced by some real and convnient path) and then

`(ccldoc::output-html *d* "/path/to/output/ccl.html" :stylesheet "ccl.css")`

**Reply by JMC-design** [reply]
Why would someone write that?  I don't understand people, or maybe I'm too simple.

**Reply by mikelevins** [reply]
Why would someone write what I wrote? To cover the possibility that folks might want to investigate how ccl generates its docs, since at least some people involved in kons-9 have a lot of experience with ccl.

Why would someone write ccldoc?

Gail wrote it because she wanted a documentation generator that the CCL maintainers would be comfortable with. They had been using docbook, which some of us knew how to configure and use, but no one particularly liked.

The CCL maintainers were sufficiently comfortable with ccldoc, so it became ccl's doc generator.

**Reply by foretspaisibles** [reply]
I just tried ccldoc with success following your indications. Skimming over the generated documentation it does not seem able to use docstrings from a Lisp system, so I understand ccldoc is a general purpose documentation system — with a very strong Lisp flavour. Am I correct or did I overlook the “docstring” part in the documentation?

**Reply by mikelevins** [reply]
I think ccldoc does not do any docstring processing. As I recall, it's essentially meant to replace docbook XML, which was the system used previously to write CCL's manual.

Docbook XML has some advantages--for example, it's a part of several industrial-strength publishing toolchains that can render large publications to a wide variety of output formats reliably and well--but those advantages do not include its source format, which is raw XML.

ccldoc's main job was to be a nicer source format than XML for a bunch of experience Lisp programmers, and I think it succeeds at that.

An arguably even nicer source format that is actually compatible with DocBook XML is Asciidoc, and it also works really well, but I'm not adcovating it because it also wouldn't provide any help with docstrings, and it also depends on a fairly elaborate toolchain that depends on either Ruby or Javascript.

I've encountered several Lisp-based doc gnerators that seemed like they did a pretty decent job. I ought to rummage through them, unless the team reaches a consensus first.

**Comment by mikelevins** [comment]
If you're going to use a documentation generator then you probably want mainly two things:

1. The generator should be able to take as input text files that are convenient and comfortable to write, and generate one or more formats that are convenient and comfortable to read. HTML is one obvious output format for reasons you've already mentioned. The most obvious input format to support is Markdown, because everyone knows it, though there are other input formats that are arguably better in one way or another (e.g. restructured text or asciidoc).
2. The generator should be able to read minimally-disruptive syntax from documentation strings to autogenerate API documentation.

One Lisp doc generator that satifies both requirements is Nicholas Hafner's Staple:

  https://github.com/Shinmera/staple

It supports input texts in a format called Markless, which is generally similar to Markdown, and it also knows how to walk ASDF projects and extract and organize docstrings.

https://shinmera.github.io/staple/staple-markless/index.html

It's not the only option, nor necessarily the best one, but in order to choose a best one, you have to decide what your criteria for "best" are. Staple at least provides the two basic features that a docgen solution should have, and therefore provides a basis for comparison. You can reasonably argue that some system is better or worse than Staple, as a way to qualify candidates.

An example of a more fully-worked-out publishing system is Matthew Butterick's Pollen:

https://github.com/mbutterick/pollen

I don't mean it as a proposed solution for kons-9, but as an example of the logical endpoint of document generators—an upper bound, if you like. I don't think you'd actually want to use it because it depends on Racket, and besides, it's probably overkill. But it illustrates some nice features that you might like to have if you could get them, such as a DSL for book design that you can conveniently and unobtrusively embed in your source text, and a rendering engine that can produce very nice camera-ready output.

The Racket docs for Pollen are here:

https://docs.racket-lang.org/pollen/

A nice example of an online book produced with it is here:

https://beautifulracket.com/

**Reply by JMC-design** [reply]
This is the big question I have for documenters starting my PDL, nested sexp's or state changing individual sexps, like that racket thing or ps or pdf...   e.g.  Though I am more interested in nice typography than long scrolling webpages.
```
(document :title "Ugh")
(chapter 1 :title  ....

(document (:title "ugh")
  (chapter 1 (:title ...

**Comment by kaveh808** [comment]
Updated the Readme with an overview/pitch of the system.

**Comment by foretspaisibles** [comment]
At Melusina we settled on gnu texinfo for the documentation generation for reasons I detail below. If you have TeX & texinfo installed you can jut try to download https://github.com/melusina-org/cl-confidence and run `development/documentation` to find the resulting PDF in `obj/makedoc/org.melusina.confidence.pdf`.

The documentation we are talking about is not arbitrary documentation, it is the reference manual of the system we are writing.  We are looking for a good system able to read all the docstrings in our code and turn it into a useful, distributable document. Tutorials and user documentation can be written with any documentation system we like, they do not require the support of a system with any Lisp specific knowledge.

What do we need in a reference manual, which is not already in the docstring, which can be accessed easily in the IDE? We found the reference manual is mostly useful for exposing the overall organisation of the software, so we are not so much interested in the docstrings themselves (we can read them in the IDE) but rather by the organisation of the document in chapters, sections, so mostly we need a hierarchical system that also allows us to decide the order of definitions and allows us to add a little introduction paragraph here and there, so the reader easily finds the important bits of information and understands how the different definitions provided by the software work together.

Advantages of texinfo / docstring.lisp:

- texinfo is a rather standard software, together with docbook and a few others, so it is easy to find examples to imitate, tips and tricks when troubleshooting, and documentation helping us to master the tool.
- texinfo has packages/ports on every system;
- texinfo output looks a bit boring but it is solid and does not call for tweaking, which is just fine for a documentation system.
- texinfo can use TeX/LaTeX so that we can add maths to our documentations, which we sometimes need to do.
- texinfo/docstrings.lisp does not impose a specific format on docstrings, so you can use the familiar terminal-oriented way to format docstrings and docstrings.lisp tries to beautify them, but the priority is set on online documentation.
- texinfo produces HTML, PDF and INFO files.
- texinfo/docstrings.lisp is super easy to setup and to use.
- texinfo can also be used to write other documents (tutorial, user manual) or to generate documentation from other systems (not a lisp-specific system)

Downsides of texinfo / docstrings.lisp

- texinfo relies on tex, a heavy software distribution (easily 3-12G, if that matters)
- texinfo relies on tex, a rather arcane software, so errors tend to be somewhat confusing to newcomers.
- the HTML produced is really plain but a good HTML/CSS artist can probably makes this beautiful
- getting the navigation structure right is easy but a little bit annoying.

**Reply by kaveh808** [reply]
I figure texinfo is the granddaddy of documentation systems, but it may be overly complex and overkill for our needs.

Last time I had to use LaTeX was for formatting my thesis, 10+ years ago, and I found it very confusing.

You mention organizing the document into chapters and sections. I find that useful.  Where and how is this text written? Is it in the source files or elsewhere?

**Reply by foretspaisibles** [reply]
In this system there is an automatism extract docstrings and generating texinfo documentation snippets out of them and we write a document including all these snippets, that document features the organization into chapters and sections.

You can see an example of such a document here: https://github.com/melusina-org/cl-confidence/blob/main/doc/org.melusina.confidence.texinfo#L79

The snippets are generated with a SBCL specific program called docstrings, which is public domain and could be used as a basis to target other systems than GNU texinfo. The docstrings program has a good understanding of Common Lisp generic methods, classes and structures and does a descent job at turning ASCII docstrings into a structured document snippet. The generated snippets look like this

~~~
% cat ~/Melusina/confidence/obj/makedoc/include/fun-org.melusina.confidence-assert-t.texinfo
@anchor{Function assert-t}
@deffn {Function} {assert-t} expr
@findex assert-t
Assert that @code{expr} is @code{t}.
This does not compare @code{expr} to @code{t} as a generalised boolean, cf. @code{assert-t*}.
@end deffn
~~~

and for a moderately more complex example:
~~~
% cat ~/Melusina/rashell/obj/makedoc/include/class-org.melusina.rashell-command.texinfo

@anchor{Class command}
@deftp {Class} {command}
@tindex command
Class precedence list: @code{@w{command}, @w{standard-object}, @w{t}}

Slots:
@itemize
@item @code{program} --- initarg: @code{@w{:program}}

A path to the program to run.
@item @code{argv} --- initarg: @code{@w{:argv}}

A sequence to be used as the argument vector for the program.
@item @code{directory} --- initarg: @code{@w{:directory}}

The working directory of the program to run.
If not provided, the current working directory is used.
@item @code{environment} --- initarg: @code{@w{:environment}}

Environment variable bindings for the program to run.
The @code{environment} must be a sequence whose terms are:
@itemize
@item  maybe the keywords @code{:append} at the first position,
   meaning the environment definitions should be appended to
   the environment of the current process.
@item  maybe the keyword @code{:supersede} at the first position, meaning
   that the environment definitions describe the entire environment
   definitions available for the external process.
@item  either a string of the form @code{"VARIABLE=VALUE"};
@item  or a cons cell of the form @code{( VARIABLE . VALUE)}.

@end itemize
When the @code{environment} is @code{nil}, then the environment of the calling process
is inherited.
@item @code{object-of-output-line} --- initarg: @code{@w{:object-of-output-line}}

When provided, a function to interpret output of the program as an object stream.
The provided function should convert a trimmed output line to the desired value.  The
conversion is only used when operation the command as a query.
@item @code{process} --- 

The external process running the program.
@item @code{documentation} --- initarg: @code{@w{:documentation}}

The documentation of the command instance.
@end itemize

The @code{command} structure captures the parameters and state of an external program.
The current state of the external program can be examined with the methods

  @code{command-status}, @code{command-input}, @code{command-output} and @code{command-error}.
@end deftp
~~~

**Comment by aartaka** [comment]
I'm going to say quite a counter-intuitive thing there (especially given @kaveh808's previous words on docstring making the code less readable): use docstrings and other built-in ways of documentations. Rationale:
- It's portable — docstrings are supported by almost all the implementations out there, and you don't have to have any dependencies or workflows to simply look at or edit your docs. It's also portable across types of things to document: functions, classes, structures, variables, macros, packages, and, I gues, some other things I've forgotten — all have their ways to define documentation, which is quite likely much more of a granularity level than any documentation system can provide 
- (Kinda follows from the previous point) It's graphics-agnostic, so you cal look at the docs in the terminal, on your phone, on the barebones SSH-ed machine with a mere Lisp compiler installed, — everywhere!
- It's conventional. There's a certain more or less universal style for docstrings that's shared by most widespread projects and that we ourselves use in Nyxt:
  - ALL-CAPS: argument of the function or slots of the class.
  - \`backtick-quote': reference to some other exported symbol from the same/other package. It's Emacs xref syntax, but, I'm guessing, it's pretty safe to use too.
- (Extremely opinionated point) It's extremely clean and readable, whe properly taken care of. If you only have plaintext docs, you are focusing on the content, not formatting.

**Reply by JMC-design** [reply]
As well, docstrings don't have to be in the code. e.g. https://github.com/JMC-design/font/blob/master/documentation.lisp
I see I didn't take away the cons dots in this one, more readable without.

**Reply by kaveh808** [reply]
Thank you for your input @aartaka. I appreciate your points, and agree that documentation should be accessible on all devices in a straightforward way.

Are there other standard or by-convention ways of embedding documentation in source files? For example, I'm interested in being able to write three or four paragraphs describing a class, its slots and methods, but not having that text interleaved in the code.

I had a look at some of the next source code, and while it does seem very well documented, it is very difficult for me to read. I accept that this is probably just a weird idiosyncrasy of mine. :-)

I imagine having a tag for some comment section at the top of the file to document and describe all the various contents of that file in one place. Some tool can then compile that text into HTML or PDF if we want.

As is my tendency, I'm looking for a minimal solution that can get us started. There are so many document generation packages out there, some quite elaborate, that I'm at a bit of a loss.

**Reply by kaveh808** [reply]
@JMC-design that is closer to what I had in mind. With even less keywords. :)

Which system is that and what does the formatted output look like?

**Reply by mikelevins** [reply]
"I had a look at some of the next source code, and while it does seem very well documented, it is very difficult for me to read. I accept that this is probably just a weird idiosyncrasy of mine. :-)"

Not entirely. I also find docstrings hard to read and offputting, and you won't find them in much of the code I write for myself.

However, in a team effort, I try to set aside such preferences in order to better serve the team's goals. If the consensus is that docstrings are the way to go, then I will support and promote that decision.

**Reply by JMC-design** [reply]
it's just that one function at the top setf'ing the docstrings.  So I guess any of the other html doc generators that scrape docstrings would work with it.  Sometimes I tweak it to document the package and symbols as well.  I too hate docstrings messing up my code flow.
it's my simple stand in until I ever get to a documentation system at the bottom of my list.

**Reply by JMC-design** [reply]
example of documenting both package and functions, throw in a :symbol in there and do symbols as well.
Still don't know why I used cons, those dots are annoying.
https://github.com/JMC-design/surface/blob/master/documentation.lisp

**Reply by kaveh808** [reply]
Thank you everyone for your comments.

It seems that docstrings are a convention I need to accept. They do serve a purpose in `describe` and for our eventual goals of an IDE.

As a convention for our project, I would like to propose the following (which we can change later if necessary):

1) At the top of the file in a comment section, we write the general description of the classes , their purpose, design decisions, and the like in plain text, using the convention of upper case for symbol names.

2) We keep docstrings compact and to the point, no more than 3-4 lines long. No blank lines. More lengthy descriptions can go at the top of the file.

3) This may be controversial, but if the code is self-documenting, skip the docstring. E.g. we don't need a docstring for `point-dot-product`. :)

Maybe we can start with this and see how things go. I need to make a pot of coffee one of these days and start by documenting the kernel files!

**Reply by aartaka** [reply]
> Are there other standard or by-convention ways of embedding documentation in source files? For example, I'm interested in being able to write three or four paragraphs describing a class, its slots and methods, but not having that text interleaved in the code.

As @JMC-design mentioned, there is: [`(setf (documentations symbol type) "...")`](http://www.lispworks.com/documentation/HyperSpec/Body/f_docume.htm). It preserves most of the docstring benefits, so it should be just fine to use, especially if you want to separate code from the docs.
